### ISSUE TITLE
App background Crash on Samsung devices with OS version 5, CleverTap Segment SDK

### ISSUE THREAD

**Issue by kunalbhise** [CleverTap/clevertap-android-sdk#37]
Device Info 
Device: Samsung 
Android OS : 5 
Type background
Clevertap SDK : com.clevertap.android:clevertap-android-sdk:3.6.1
Segment integration : com.clevertap.android:clevertap-segment-android:+

Stacktrace : 
Fatal Exception: java.lang.ArrayIndexOutOfBoundsException: length=0; index=42375
       at java.lang.reflect.ArtField.getNameNative(ArtField.java)
       at java.lang.reflect.ArtField.getName + 91(ArtField.java:91)
       at java.lang.reflect.Field.getName + 122(Field.java:122)
       at com.clevertap.android.sdk.ab_testing.uieditor.ResourceIds.readClassIds + 85(ResourceIds.java:85)
       at com.clevertap.android.sdk.ab_testing.uieditor.ResourceIds.read + 34(ResourceIds.java:34)
       at com.clevertap.android.sdk.ab_testing.uieditor.ResourceIds.<init> + 21(ResourceIds.java:21)
       at com.clevertap.android.sdk.ab_testing.uieditor.UIEditor.<init> + 165(UIEditor.java:165)
       at com.clevertap.android.sdk.ab_testing.CTABTestController.<init> + 197(CTABTestController.java:197)
       at com.clevertap.android.sdk.CleverTapAPI.initABTesting + 7212(CleverTapAPI.java:7212)
       at com.clevertap.android.sdk.CleverTapAPI.<init> + 155(CleverTapAPI.java:155)
       at com.clevertap.android.sdk.CleverTapAPI.instanceWithConfig + 509(CleverTapAPI.java:509)
       at com.clevertap.android.sdk.CleverTapAPI.getDefaultInstance + 460(CleverTapAPI.java:460)
       at com.clevertap.android.sdk.CleverTapAPI.getDefaultInstance + 475(CleverTapAPI.java:475)
       at com.segment.analytics.android.integrations.clevertap.CleverTapIntegration$1.create + 77(CleverTapIntegration.java:77)
       at com.segment.analytics.Analytics.performInitializeIntegrations + 1405(Analytics.java:1405)
       at com.segment.analytics.Analytics$2$1.run + 277(Analytics.java:277)
       at android.os.Handler.handleCallback + 739(Handler.java:739)
       at android.os.Handler.dispatchMessage + 95(Handler.java:95)
       at android.os.Looper.loop + 145(Looper.java:145)
       at android.app.ActivityThread.main + 6946(ActivityThread.java:6946)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke + 372(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run + 1404(ZygoteInit.java:1404)
       at com.android.internal.os.ZygoteInit.main + 1199(ZygoteInit.java:1199)

**Comment by darshanclevertap** [comment]
@kunalbhise Thank you for bringing this to our attention.
Could you help us with a few details?

1. Is this happening only on Android OS version 5 or for other OS versions as well?
2. Is this happening only on Samsung devices or other devices as well?
3. Does this crash happen on the background silently or it affects the app UI ?

These details will help us debug the issue better.
Thanks!

**Comment by kunalbhise** [comment]
1. this happening only on Android OS version 5 or for other OS versions as well? 
- Yes
2. Is this happening only on Samsung devices or other devices as well?
- No didn't find trace for  other devices
3. Does this crash happen on the background silently or it affects the app UI ?
  - As per Crashlytics data its happening in background. Please find screenshot below.

<img width="1041" alt="Screenshot 2019-11-11 at 6 29 37 PM" src="https://user-images.githubusercontent.com/13194070/68589287-85478200-04b1-11ea-90c4-d7d9bf839b3a.png">

Thanks!

**Comment by darshanclevertap** [comment]
@kunalbhise the stack trace is collapsed in the screenshot. If you could share the entire stack trace here, it would be helpful. Also, I didn't understand your answer to the first question, is it "yes, it is happening for only Android 5" or is it "yes, it is happening for all the versions"?

**Comment by kunalbhise** [comment]
@darshanclevertap 
Please find the Attached txt for stack trace.
[clevertap_issue_crash_5DD4B4EC03980001378B96B2CCB4C9BA_DNE_0_v2.txt](https://github.com/CleverTap/clevertap-android-sdk/files/3867472/clevertap_issue_crash_5DD4B4EC03980001378B96B2CCB4C9BA_DNE_0_v2.txt)

Recently I have observed crash on Android version 7 but the majority of the crashes are on Android 5. please see below screenshot.

<img width="1016" alt="Screenshot 2019-11-20 at 11 14 46 AM" src="https://user-images.githubusercontent.com/13194070/69212340-0aecc100-0b87-11ea-99fa-783178c7a7ca.png">

**Comment by darshanclevertap** [comment]
@kunalbhise Thank you for the logs, looks like something is up with Android Lollipop OS code so give me some time to look into this from my end and I will keep you updated.

**Comment by kumaratulismu** [comment]
Closing due to inactivity.

**Comment by ParveenDala** [comment]
Hi @darshanclevertap, Any update on this?
We observed the same issue with CleverTap **3.8.0**
<img width="437" alt="Screenshot 2020-07-15 at 1 39 15 PM" src="https://user-images.githubusercontent.com/31188015/87520442-9aa82800-c6a0-11ea-9b7f-872ebccdac38.png">

**Comment by darshanclevertap** [comment]
@ParveenDala Can you share the stack trace?

**Comment by ParveenDala** [comment]
```
Fatal Exception: java.lang.ArrayIndexOutOfBoundsException
<muted>
Fatal Exception: java.lang.ArrayIndexOutOfBoundsException: <muted>
       at java.lang.reflect.ArtField.getNameNative(ArtField.java)
       at java.lang.reflect.ArtField.getName(ArtField.java:91)
       at java.lang.reflect.Field.getName(Field.java:122)
       at com.clevertap.android.sdk.ab_testing.uieditor.ResourceIds.readClassIds(ResourceIds.java:85)
       at com.clevertap.android.sdk.ab_testing.uieditor.ResourceIds.read(ResourceIds.java:34)
       at com.clevertap.android.sdk.ab_testing.uieditor.ResourceIds.<init>(ResourceIds.java:21)
       at com.clevertap.android.sdk.ab_testing.uieditor.UIEditor.<init>(UIEditor.java:165)
       at com.clevertap.android.sdk.ab_testing.CTABTestController.<init>(CTABTestController.java:197)
       at com.clevertap.android.sdk.CleverTapAPI.initABTesting(CleverTapAPI.java:7871)
       at com.clevertap.android.sdk.CleverTapAPI.<init>(CleverTapAPI.java:179)
       at com.clevertap.android.sdk.CleverTapAPI.instanceWithConfig(CleverTapAPI.java:551)
       at com.clevertap.android.sdk.CleverTapAPI.getDefaultInstance(CleverTapAPI.java:502)
       at com.clevertap.android.sdk.CleverTapAPI.getDefaultInstance(CleverTapAPI.java:517)
```
Happens when call: CleverTapAPI.getDefaultInstance(context)
Please find the attached file for detailed logs:
[CleverTapIssueInit.txt](https://github.com/CleverTap/clevertap-android-sdk/files/4925126/CleverTapIssueInit.txt)

**Comment by ParveenDala** [comment]
@darshanclevertap 
```
Fatal Exception: java.lang.ArrayIndexOutOfBoundsException
<muted>
Fatal Exception: java.lang.ArrayIndexOutOfBoundsException: <muted>
       at java.lang.reflect.ArtField.getNameNative(ArtField.java)
       at java.lang.reflect.ArtField.getName(ArtField.java:91)
       at java.lang.reflect.Field.getName(Field.java:122)
       at com.clevertap.android.sdk.ab_testing.uieditor.ResourceIds.readClassIds(ResourceIds.java:85)
       at com.clevertap.android.sdk.ab_testing.uieditor.ResourceIds.read(ResourceIds.java:34)
       at com.clevertap.android.sdk.ab_testing.uieditor.ResourceIds.<init>(ResourceIds.java:21)
       at com.clevertap.android.sdk.ab_testing.uieditor.UIEditor.<init>(UIEditor.java:165)
       at com.clevertap.android.sdk.ab_testing.CTABTestController.<init>(CTABTestController.java:197)
       at com.clevertap.android.sdk.CleverTapAPI.initABTesting(CleverTapAPI.java:7871)
       at com.clevertap.android.sdk.CleverTapAPI.<init>(CleverTapAPI.java:179)
       at com.clevertap.android.sdk.CleverTapAPI.instanceWithConfig(CleverTapAPI.java:551)
       at com.clevertap.android.sdk.CleverTapAPI.getDefaultInstance(CleverTapAPI.java:502)
       at com.clevertap.android.sdk.CleverTapAPI.getDefaultInstance(CleverTapAPI.java:517)
```

Happens when call: CleverTapAPI.getDefaultInstance(context)
Please find the attached file for detailed logs:
[CleverTapIssueInit.txt](https://github.com/CleverTap/clevertap-android-sdk/files/4925254/CleverTapIssueInit.txt)

**Comment by vamshirupeek** [comment]
@darshanclevertap 

Facing similar issue but couldn't find the root cause for this issue.


```
com.clevertap.android.sdk.ab_testing.CTABTestController
       at android.os.MessageQueue.nativePollOnce(MessageQueue.java)
       at android.os.MessageQueue.next(MessageQueue.java:143)
       at android.os.Looper.loop(Looper.java:130)
       at android.os.HandlerThread.run(HandlerThread.java:61)
```

**Comment by vamshirupeek** [comment]
Facing another such issue. 

<img width="439" alt="Screenshot 2020-08-24 at 2 01 40 PM" src="https://user-images.githubusercontent.com/45193823/91022183-6c5c2780-e612-11ea-82aa-9eb04db77e76.png">


```
Fatal Exception: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.clevertap.android.sdk.CleverTapAPI.setLibrary(java.lang.String)' on a null object reference
       at com.segment.analytics.android.integrations.clevertap.CleverTapIntegration.<init>(CleverTapIntegration.java:96)
       at com.segment.analytics.android.integrations.clevertap.CleverTapIntegration$1.create(CleverTapIntegration.java:81)
       at com.segment.analytics.Analytics.performInitializeIntegrations(Analytics.java:1436)
       at com.segment.analytics.Analytics$2$1.run(Analytics.java:279)
       at android.os.Handler.handleCallback(Handler.java:883)
       at android.os.Handler.dispatchMessage(Handler.java:100)
       at android.os.Looper.loop(Looper.java:237)
       at android.app.ActivityThread.main(ActivityThread.java:7948)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1075)
```

**Comment by darshanclevertap** [comment]
@ParveenDala Our next version of the Android SDK will fix this issue.

@vamshirupeek You've posted the stack trace of our Segment SDK, once we release the next Android SDK with the fix, we will update the Segment SDK as well.

This issue will be open till then. We will post here once the release is done.
Thank you for your patience! 🙇

**Comment by singh496** [comment]
can this issue also cause an ANR? We're facing anr's on our dashboard and it looks like one the threads running has similar output as to [this](https://github.com/CleverTap/clevertap-android-sdk/issues/37#issuecomment-678983508)

**Comment by darshanclevertap** [comment]
@singh496 We've already added a fix for this which will be released tomorrow, you ideally shouldn't be getting ANRs, can you create a new issue with the ANR stack trace?

**Comment by darshanclevertap** [comment]
@vamshirupeek @kunalbhise We've fixed this in v3.9.0. We will update our Segment SDK as well and give you an update.

**Comment by darshanclevertap** [comment]
@kunalbhise @ParveenDala We've released the CleverTap Segment SDK v1.1.9 with this fix. Please upgrade to the latest SDK. Thanks!

**Comment by vamshirupeek** [comment]
@darshanclevertap We have updated to the latest versions of both CleverTap SDK and Segment SDK but still we are facing the background crashes can we know what is the reason why is it still happening?

**Comment by kumaratulismu** [comment]
@vamshirupeek Can you share the exact version details of CleverTap Android & Segment SDK's. There are multiple logs in this thread. Can you help us with the exact stack trace?

**Comment by vamshirupeek** [comment]
@kumaratulismu 

Here are details

CleverTap Android SDK - 3.9.0
Segment SDK - 1.1.9

Check the stacktrace below

`Caused by java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@18bd69f6 rejected from java.util.concurrent.ScheduledThreadPoolExecutor@293341f7[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
       at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2011)
       at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:793)
       at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:298)
       at java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:503)
       at java.util.concurrent.Executors$DelegatedScheduledExecutorService.schedule(Executors.java:644)
       at androidx.work.impl.background.systemalarm.WorkTimer.startTimer(WorkTimer.java:82)
       at androidx.work.impl.background.systemalarm.DelayMetCommandHandler.onAllConstraintsMet(DelayMetCommandHandler.java:100)
       at androidx.work.impl.constraints.WorkConstraintsTracker.onConstraintMet(WorkConstraintsTracker.java:150)
       at androidx.work.impl.constraints.controllers.ConstraintController.updateCallback(ConstraintController.java:134)
       at androidx.work.impl.constraints.controllers.ConstraintController.onConstraintChanged(ConstraintController.java:141)
       at androidx.work.impl.constraints.trackers.ConstraintTracker.setState(ConstraintTracker.java:103)
       at androidx.work.impl.constraints.trackers.NetworkStateTracker$NetworkStateBroadcastReceiver.onReceive(NetworkStateTracker.java:170)
       at android.app.LoadedApk$ReceiverDispatcher$Args.run(LoadedApk.java:933)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:145)
       at android.app.ActivityThread.main(ActivityThread.java:6145)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)`

**Comment by kumaratulismu** [comment]
@vamshirupeek I can't find the Clevertap classes in the stack trace. Can you please check again?

**Comment by vamshirupeek** [comment]
@kumaratulismu Please check the attached image which show the threads which are effected and the number of crashes in the last 90 days.

![screencapture-console-firebase-google-project-insight-38a5e-crashlytics-app-android-com-rupeek-customer-issues-f95f9b26ce36516e5d84540066d23633-2020-09-22-12_50_48](https://user-images.githubusercontent.com/45193823/93854210-edbade80-fcd2-11ea-951c-ab07ab4bf091.png)

**Comment by piyush-kukadiya** [comment]
@vamshirupeek stack-trace in your image shows that it's crashing because of work manager issue [here](https://github.com/googlecodelabs/android-workmanager/issues/130) and [here](https://issuetracker.google.com/issues/129226383). CleverTap threads are affected during that crash thats why its in stack-trace. Nothing to do with CleverTap crash, CleverTap is not using WorkManager in SDK.
